File: macos-widget/PlaneTrackerWidget/Services/RouteService.swift
import Foundation

enum RouteService {
    /// Fetch route info for a callsign from adsbdb
    static func fetchRoute(callsign: String) async -> RouteInfo? {
        let cleaned = callsign.trimmingCharacters(in: .whitespaces)
        guard !cleaned.isEmpty,
              let encoded = cleaned.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed),
              let url = URL(string: "\(PTConstants.adsbdbBase)/callsign/\(encoded)")
        else { return nil }

        do {
            var request = URLRequest(url: url)
            request.timeoutInterval = 10
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            let decoded = try JSONDecoder().decode(ADSBDBCallsignResponse.self, from: data)
            guard let fr = decoded.response?.flightroute else { return nil }
            return RouteInfo.from(fr)
        } catch {
            return nil
        }
    }

    /// Fetch routes for multiple callsigns concurrently
    static func fetchRoutes(for flights: [Flight]) async -> [String: RouteInfo] {
        var result: [String: RouteInfo] = [:]

        await withTaskGroup(of: (String, RouteInfo?).self) { group in
            for flight in flights {
                group.addTask {
                    let route = await fetchRoute(callsign: flight.callsign)
                    return (flight.callsign, route)
                }
            }
            for await (callsign, route) in group {
                if let route { result[callsign] = route }
            }
        }

        return result
    }
}

File: android-widget/app/src/main/java/com/planetracker/data/service/RouteService.kt
package com.planetracker.data.service

import com.planetracker.data.model.ADSBDBCallsignResponse
import com.planetracker.data.model.Flight
import com.planetracker.data.model.RouteInfo
import com.planetracker.util.Constants
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

object RouteService {
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private val moshi = Moshi.Builder()
        .addLast(KotlinJsonAdapterFactory())
        .build()

    private val responseAdapter = moshi.adapter(ADSBDBCallsignResponse::class.java)

    /** Fetch route info for a callsign from adsbdb */
    suspend fun fetchRoute(callsign: String): RouteInfo? = withContext(Dispatchers.IO) {
        val cleaned = callsign.trim()
        if (cleaned.isEmpty()) return@withContext null

        try {
            val encoded = URLEncoder.encode(cleaned, "UTF-8")
            val url = "${Constants.ADSBDB_BASE}/callsign/$encoded"
            val request = Request.Builder().url(url).build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) return@withContext null

            val body = response.body?.string() ?: return@withContext null
            val parsed = responseAdapter.fromJson(body) ?: return@withContext null
            val fr = parsed.response?.flightroute ?: return@withContext null
            RouteInfo.from(fr)
        } catch (_: Exception) {
            null
        }
    }

    /** Fetch routes for multiple flights concurrently */
    suspend fun fetchRoutes(flights: List<Flight>): Map<String, RouteInfo> =
        coroutineScope {
            flights.map { flight ->
                async {
                    val route = fetchRoute(flight.callsign)
                    if (route != null) flight.callsign to route else null
                }
            }
                .awaitAll()
                .filterNotNull()
                .toMap()
        }
}
